src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -381,34 +489,59 @@ static void dds_stream_countops1 (const uint32_t * __restrict ops, const uint32_
         ops++;
         break;
       }
-      case DDS_OP_RTS: case DDS_OP_JEQ: {
+      case DDS_OP_RTS: case DDS_OP_JEQ: case DDS_OP_KOF: {
         abort ();
         break;
       }
+      case DDS_OP_DLC: {
+        ops++;
+        break;
+      }
+      case DDS_OP_PLC: {
+        ops = dds_stream_countops_pl (ops, ops_end);
+        break;
+      }
     }
   }
   ++ops; /* skip RTS op */
   if (ops > *ops_end)
     *ops_end = ops;
 }
 
-uint32_t dds_stream_countops (const uint32_t * __restrict ops)
+static void dds_stream_countops_keyoffset (const uint32_t * __restrict ops, const dds_key_descriptor_t * __restrict key, const uint32_t ** __restrict ops_end)
+{
+  assert (key);
+  assert (*ops_end);
+  if (key->m_index >= *ops_end - ops)
+  {
+    assert (DDS_OP (ops[key->m_index]) == DDS_OP_KOF);
+    *ops_end = ops + key->m_index + 1 + DDS_OP_LENGTH (ops[key->m_index]);
+  }
+}
+
+uint32_t dds_stream_countops (const uint32_t * __restrict ops, uint32_t nkeys, const dds_key_descriptor_t * __restrict keys)
 {
   const uint32_t *ops_end = ops;
   dds_stream_countops1 (ops, &ops_end);
+  for (uint32_t n = 0; n < nkeys; n++)
+    dds_stream_countops_keyoffset (ops, &keys[n], &ops_end);
   return (uint32_t) (ops_end - ops);
 }
 
-static void dds_stream_reuse_string_bound (dds_istream_t * __restrict is, char * __restrict str, const uint32_t bound)
+static char *dds_stream_reuse_string_bound (dds_istream_t * __restrict is, char * __restrict str, const uint32_t bound, bool alloc)
 {
   const uint32_t length = dds_is_get4 (is);
   const void *src = is->m_buffer + is->m_index;
   /* FIXME: validation now rejects data containing an oversize bounded string,
      so this check is superfluous, but perhaps rejecting such a sample is the
      wrong thing to do */
-  assert (str != NULL);
+  if (!alloc)
+    assert (str != NULL);
+  else if (str == NULL)
+    str = dds_alloc (bound);
```

Comment:
Bound also included the terminating 0 (see the description in `dds_opcodes.h`), so that means the name `bound` is confusing here. I've replaced it by `size` (also in the opcodes header file).